src/translation.cpp

```diff
@@ -143,14 +220,338 @@ void Translations::loadTranslation(const std::string &data)
 			}
 		}
 
-		std::wstring oword1 = word1.str(), oword2 = word2.str();
-		if (!oword2.empty()) {
-			std::wstring translation_index = textdomain + L"|";
-			translation_index.append(oword1);
-			m_translations[translation_index] = oword2;
-		} else {
-			infostream << "Ignoring empty translation for \""
-				<< wide_to_utf8(oword1) << "\"" << std::endl;
+		addTranslation(textdomain, word1.str(), word2.str());
+	}
+}
+
+
+std::wstring Translations::unescapeC(const std::wstring &str) {
+	// Process escape sequences in str as if it were a C string
+	std::wstring result;
+	size_t i = 0;
+	while (i < str.length()) {
+		if (str[i] != L'\\') {
+			result.push_back(str[i]);
+			i++;
+			continue;
+		}
+		i++;
+		if (i == str.length()) {
+			errorstream << "Unfinished escape sequence at the end of \"" << wide_to_utf8(str) << "\"" << std::endl;
+			break;
+		}
+		switch (str[i]) {
+			// From https://en.wikipedia.org/wiki/Escape_sequences_in_C#Table_of_escape_sequences
+			case L'a': result.push_back(L'\a'); break;
+			case L'b': result.push_back(L'\b'); break;
+			case L'e': result.push_back(L'\x1b'); break;
+			case L'f': result.push_back(L'\f'); break;
+			case L'n': result.push_back(L'\n'); break;
+			case L'r': result.push_back(L'\r'); break;
+			case L't': result.push_back(L'\t'); break;
+			case L'v': result.push_back(L'\v'); break;
+			case L'\\': result.push_back(L'\\'); break;
+			case L'\'': result.push_back(L'\''); break;
+			case L'"': result.push_back(L'"'); break;
+			case L'?': result.push_back(L'?'); break;
+			case L'0': case L'1': case L'2': case L'3': case L'4': case L'5': case L'6': case L'7': {
+				size_t j = 0;
+				wchar_t c = 0;
+				for (; j < 3 && i+j < str.length() && L'0' <= str[i+j] && str[i+j] <= L'7'; j++) {
+					c = c * 8 + (str[i+j] - L'0');
+				}
+				if (c <= 0xff) {
+					result.push_back(c);
+				}
+				i += j;
+				continue;
+			}
+			case L'x': {
+				i++;
+				if (i >= str.length()) {
+					errorstream << "Unfinished escape sequence at the end of \"" << wide_to_utf8(str) << "\"" << std::endl;
+				}
+				char32_t c = 0;
+				size_t j = 0;
+				unsigned char v;
+				for (; i+j < str.length() && hex_digit_decode((char)str[i+j], v); j++) {
```

Comment:
Do I see it correctly that `\x` takes an arbitrary number of hex characters?
Shouldn't that be exactly two?